src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataFilter.cs

```diff
@@ -66,6 +74,14 @@ public void OnResourceExecuting(ResourceExecutingContext context)
         /// <inheritdoc />
         public void OnResourceExecuted(ResourceExecutedContext context)
         {
+            // If there is an unhandled exception, we would like to avoid setting tempdata as 
+            // the end user is going to see an error page anyway and also it helps us in avoiding
+            // accessing resources like Session too late in the request lifecyle where SessionFeature might
+            // not be available.
+            if (!context.HttpContext.Response.HasStarted && context.Exception != null)
+            {
+                context.HttpContext.Items.Add(TempDataUnhandledExceptionKey, true);
+            }
```

Comment:
This is kinda a best-effort mechanism. This is likely to help in a lot of cases, but it's not airtight.